host/programexecutor.go

```diff
@@ -0,0 +1,665 @@
+	if err := validateRegistryEntry(value); err != nil {
+		return fmt.Errorf("invalid registry value: %w", err)
+	}
+
+	key := value.Key()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	if err := ex.host.lockRegistryKey(ctx, key); err != nil {
+		return fmt.Errorf("failed to lock registry key: %w", err)
+	}
+	defer ex.host.unlockRegistryKey(key)
+
+	expirationHeight := ex.vc.Index.Height + 52560
+
+	// get the current value.
+	oldValue, err := ex.host.registry.Get(key)
+	// if the key doesn't exist, we don't need to validate it further.
+	if errors.Is(err, ErrEntryNotFound) {
+		if _, err = ex.host.registry.Set(key, value, expirationHeight); err != nil {
+			return fmt.Errorf("failed to create registry key: %w", err)
+		}
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to get registry value: %w", err)
+	}
+
+	h := types.NewHasher()
+	ex.hostKey.EncodeTo(h.E)
+	hostID := h.Sum()
+
+	// validate that the update is valid.
+	if err := validateRegistryUpdate(oldValue, value, hostID); err != nil {
+		return fmt.Errorf("invalid registry update: %w", err)
+	}
+
+	// set the new value.
+	if _, err = ex.host.registry.Set(key, value, expirationHeight); err != nil {
+		err = fmt.Errorf("failed to update registry value: %w", err)
+	}
+	return err
+}
+
+// ExecuteInstruction executes the given instruction, reading program data from r
+// as needed, and writing the result of the instruction to w.
+//
+// note: Unlike siad's MDM, this implementation does not check the
+// data offsets in the instruction arguments. It is assumed the program data is
+// well-formed, meaning each argument appears in the program data in the order it
+// is needed. Malformed programs may lead to unexpected behavior, but there is
+// no need to buffer the program's data in memory during execution. Changing an
+// instructions arguments would also cause programs to be unexecutable, so this
+// seems like an acceptable trade-off. Should consider removing the offsets
+// from the instruction arguments.
+func (ex *programExecutor) ExecuteInstruction(r io.Reader, w io.Writer, instruction rhp.Instruction) error {
+	if ex.committed {
+		panic("cannot modify a committed program")
+	}
+
+	var proof []types.Hash256
+	var err error
+
+	ex.output.Reset()
+
+	switch instruction.Specifier {
+	case rhp.InstrAppendSector:
+		if len(instruction.Arguments) != rhp.LenAppendSectorArgs {
+			err = errInvalidArgs
+			break
+		}
+
+		// read the sector data.
+		var sector [rhp.SectorSize]byte
+		if _, err = io.ReadFull(r, sector[:]); err != nil {
+			err = fmt.Errorf("failed to read append sector data: %w", err)
+			break
+		}
+
+		proof, err = ex.executeAppendSector(&sector, instruction.Arguments[8] == 1)
+	case rhp.InstrDropSectors:
+		if len(instruction.Arguments) != rhp.LenDropSectorsArgs {
+			err = errInvalidArgs
+			break
+		}
+
+		var dropped uint64
+		if err = binary.Read(r, binary.LittleEndian, &dropped); err != nil {
```

Comment:
Because arguments for drop sectors in the RHP3 MDM is a uint64 data offset to the uint64 number of sectors to drop.

If we're not worried about compatibility, it might be worth ditching the arguments/data distinction and moving to something closer to the RHP2 RPC loop. A loop of 16-byte op-code, uint64 data length, then the data to execute the instruction, a `done` op-code would signal finalization.